File: src/222. Count Complete Tree Nodes.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

/**
 * @param {TreeNode} root
 * @return {number}
 */

/*
function countNodes (root) {
  return root
    ? 1 + countNodes(root.left) + countNodes(root.right)
    : 0
}
 */

function countNodes(root) {
  let hLeft = 0,
    hRight = 0
  let pLeft = root,
    pRight = root
  while (pLeft) {
    hLeft++
    pLeft = pLeft.left
  }
  while (pRight) {
    hRight++
    pRight = pRight.right
  }
  return hLeft === hRight
    ? 2 ** hLeft - 1
    : 1 + countNodes(root.left) + countNodes(root.right)
}

File: src/222. Count Complete Tree Nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int countNodes(TreeNode* root) {
        int hLeft = 0, hRight = 0;
        TreeNode *pLeft = root, *pRight = root;

        while (pLeft) {
            hLeft++;
            pLeft = pLeft->left;
        }

        while (pRight) {
            hRight++;
            pRight = pRight->right;
        }

        return hLeft == hRight
            ? pow(2, hLeft) - 1
            : 1 + countNodes(root->left) + countNodes(root->right);
        }
};